### ISSUE TITLE
bun-types: doesn't regcognise methods `json`, `text`, etc. on `BunFile` instances

### ISSUE THREAD

**Issue by maranomynet** [oven-sh/bun#8103]
### What version of Bun is running?

`1.0.22`   (with `@types/bun@1.0.1`)

### What platform is your computer?

Darwin 23.1.0 arm64 arm

### What steps can reproduce the bug?

`test.ts`:
```ts
console.log(await Bun.file('./foo.json').json());
```

### What is the expected behavior?

No type error.

### What do you see instead?

```
Property 'json' does not exist on type 'BunFile'. ts(2339)
```

<img width="746" alt="image" src="https://github.com/oven-sh/bun/assets/86827/d9ab8249-1181-4599-9917-a499750f3bc5">

### Additional information

If I pin `@types/bun` at version `1.0.0`, this type error disappears.

I have tested the above with both TypeScript `^4.9`  and `^5.3`

**Comment by asomethings** [comment]
I believe you have `"DOM"` in `tsconfig.json`s `lib` configuration. 

Bun's `Blob` will only infer type on below condition

```typescript
type _Blob = typeof globalThis extends { onerror: any; Blob: infer T }
  ? T
  : Blob;
```

**Comment by maranomynet** [comment]
> I believe you have `"DOM"` in `tsconfig.json`s `lib` configuration.

I don't think so. Here's my `tsconfig.json`:

```jsonc
{
  "compilerOptions": {
    "lib": ["esnext"],
    "allowSyntheticDefaultImports": true,
    "module": "NodeNext",
    "target": "ES2017",
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "noEmit": true,
    "jsx": "react",

    "strict": true.
    "noUncheckedIndexedAccess": true,
    "noImplicitAny": true,
    "useUnknownInCatchVariables": true
  },
  // include and exclude ...
}
```

**Comment by asomethings** [comment]
Hmm... that's interesting... 

It seems it inferences type for example below. Could you provide an reproducible repository?


<img width="1277" src="https://github.com/oven-sh/bun/assets/16171942/ddd8ed6a-f77d-4876-859a-05bea0b155a0">

**Comment by maranomynet** [comment]
@asomethings Thanks for looking into this.
I had this consistently going on in two different repositories, but was unable to replicate the issue in a freshly generated repo.

I then managed to resolve the problem in both repos by doing `rm -rf node_modules bun.lockb && bun install` 

The two problem repos had one thing in common: They both got converted to bun, over a pre-existing node_modules folder.

I guess it's imperative to start any conversion to bun with a quick `rm -rf node_modules`.

**Comment by warcayac** [comment]
I recently upgraded bun to 1.0.22 and created a fresh server with
```
$ bun create elysia phonebook_server
```

**index.ts**
```typescript
import Elysia, {t} from "elysia";

const file = Bun.file('./data/persons.json');
const DB_PERSONS = await file.json();

const app = new Elysia()
  .get(
    "/",
    () => "Hello Elysia",
  )
  .listen(3000)
;

console.log(
  `🦊 Elysia is running at ${app.server?.hostname}:${app.server?.port}`
);
```
And I'm getting this error:
```
Property 'json' does not exist on type 'BunFile'.ts(2339)
```
No modification has been made to other files. 

I've tried the solution of @maranomynet 

> I then managed to resolve the problem in both repos by doing rm -rf node_modules bun.lockb && bun install

but the problem persists.

**package.json**
```
{
  "name": "phonebook_server",
  "version": "1.0.50",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "dev": "bun run --watch src/index.ts"
  },
  "dependencies": {
    "elysia": "latest"
  },
  "devDependencies": {
    "bun-types": "latest"
  },
  "module": "src/index.js"
}
```

**tsconfig.json**
```
{
  "compilerOptions": {
    /* Visit https://aka.ms/tsconfig to read more about this file */

    /* Projects */
    // "incremental": true,                              /* Save .tsbuildinfo files to allow for incremental compilation of projects. */
    // "composite": true,                                /* Enable constraints that allow a TypeScript project to be used with project references. */
    // "tsBuildInfoFile": "./.tsbuildinfo",              /* Specify the path to .tsbuildinfo incremental compilation file. */
    // "disableSourceOfProjectReferenceRedirect": true,  /* Disable preferring source files instead of declaration files when referencing composite projects. */
    // "disableSolutionSearching": true,                 /* Opt a project out of multi-project reference checking when editing. */
    // "disableReferencedProjectLoad": true,             /* Reduce the number of projects loaded automatically by TypeScript. */

    /* Language and Environment */
    "target": "ES2021",                                  /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */
    // "lib": [],                                        /* Specify a set of bundled library declaration files that describe the target runtime environment. */
    // "jsx": "preserve",                                /* Specify what JSX code is generated. */
    // "experimentalDecorators": true,                   /* Enable experimental support for TC39 stage 2 draft decorators. */
    // "emitDecoratorMetadata": true,                    /* Emit design-type metadata for decorated declarations in source files. */
    // "jsxFactory": "",                                 /* Specify the JSX factory function used when targeting React JSX emit, e.g. 'React.createElement' or 'h'. */
    // "jsxFragmentFactory": "",                         /* Specify the JSX Fragment reference used for fragments when targeting React JSX emit e.g. 'React.Fragment' or 'Fragment'. */
    // "jsxImportSource": "",                            /* Specify module specifier used to import the JSX factory functions when using 'jsx: react-jsx*'. */
    // "reactNamespace": "",                             /* Specify the object invoked for 'createElement'. This only applies when targeting 'react' JSX emit. */
    // "noLib": true,                                    /* Disable including any library files, including the default lib.d.ts. */
    // "useDefineForClassFields": true,                  /* Emit ECMAScript-standard-compliant class fields. */
    // "moduleDetection": "auto",                        /* Control what method is used to detect module-format JS files. */

    /* Modules */
    "module": "ES2022",                                /* Specify what module code is generated. */
    // "rootDir": "./",                                  /* Specify the root folder within your source files. */
    "moduleResolution": "node",                       /* Specify how TypeScript looks up a file from a given module specifier. */
    // "baseUrl": "./",                                  /* Specify the base directory to resolve non-relative module names. */
    // "paths": {},                                      /* Specify a set of entries that re-map imports to additional lookup locations. */
    // "rootDirs": [],                                   /* Allow multiple folders to be treated as one when resolving modules. */
    // "typeRoots": [],                                  /* Specify multiple folders that act like './node_modules/@types'. */
    "types": ["bun-types"],                                      /* Specify type package names to be included without being referenced in a source file. */
    // "allowUmdGlobalAccess": true,                     /* Allow accessing UMD globals from modules. */
    // "moduleSuffixes": [],                             /* List of file name suffixes to search when resolving a module. */
    // "resolveJsonModule": true,                        /* Enable importing .json files. */
    // "noResolve": true,                                /* Disallow 'import's, 'require's or '<reference>'s from expanding the number of files TypeScript should add to a project. */

    /* JavaScript Support */
    // "allowJs": true,                                  /* Allow JavaScript files to be a part of your program. Use the 'checkJS' option to get errors from these files. */
    // "checkJs": true,                                  /* Enable error reporting in type-checked JavaScript files. */
    // "maxNodeModuleJsDepth": 1,                        /* Specify the maximum folder depth used for checking JavaScript files from 'node_modules'. Only applicable with 'allowJs'. */

    /* Emit */
    // "declaration": true,                              /* Generate .d.ts files from TypeScript and JavaScript files in your project. */
    // "declarationMap": true,                           /* Create sourcemaps for d.ts files. */
    // "emitDeclarationOnly": true,                      /* Only output d.ts files and not JavaScript files. */
    // "sourceMap": true,                                /* Create source map files for emitted JavaScript files. */
    // "outFile": "./",                                  /* Specify a file that bundles all outputs into one JavaScript file. If 'declaration' is true, also designates a file that bundles all .d.ts output. */
    // "outDir": "./",                                   /* Specify an output folder for all emitted files. */
    // "removeComments": true,                           /* Disable emitting comments. */
    // "noEmit": true,                                   /* Disable emitting files from a compilation. */
    // "importHelpers": true,                            /* Allow importing helper functions from tslib once per project, instead of including them per-file. */
    // "importsNotUsedAsValues": "remove",               /* Specify emit/checking behavior for imports that are only used for types. */
    // "downlevelIteration": true,                       /* Emit more compliant, but verbose and less performant JavaScript for iteration. */
    // "sourceRoot": "",                                 /* Specify the root path for debuggers to find the reference source code. */
    // "mapRoot": "",                                    /* Specify the location where debugger should locate map files instead of generated locations. */
    // "inlineSourceMap": true,                          /* Include sourcemap files inside the emitted JavaScript. */
    // "inlineSources": true,                            /* Include source code in the sourcemaps inside the emitted JavaScript. */
    // "emitBOM": true,                                  /* Emit a UTF-8 Byte Order Mark (BOM) in the beginning of output files. */
    // "newLine": "crlf",                                /* Set the newline character for emitting files. */
    // "stripInternal": true,                            /* Disable emitting declarations that have '@internal' in their JSDoc comments. */
    // "noEmitHelpers": true,                            /* Disable generating custom helper functions like '__extends' in compiled output. */
    // "noEmitOnError": true,                            /* Disable emitting files if any type checking errors are reported. */
    // "preserveConstEnums": true,                       /* Disable erasing 'const enum' declarations in generated code. */
    // "declarationDir": "./",                           /* Specify the output directory for generated declaration files. */
    // "preserveValueImports": true,                     /* Preserve unused imported values in the JavaScript output that would otherwise be removed. */

    /* Interop Constraints */
    // "isolatedModules": true,                          /* Ensure that each file can be safely transpiled without relying on other imports. */
    // "allowSyntheticDefaultImports": true,             /* Allow 'import x from y' when a module doesn't have a default export. */
    "esModuleInterop": true,                             /* Emit additional JavaScript to ease support for importing CommonJS modules. This enables 'allowSyntheticDefaultImports' for type compatibility. */
    // "preserveSymlinks": true,                         /* Disable resolving symlinks to their realpath. This correlates to the same flag in node. */
    "forceConsistentCasingInFileNames": true,            /* Ensure that casing is correct in imports. */

    /* Type Checking */
    "strict": true,                                      /* Enable all strict type-checking options. */
    // "noImplicitAny": true,                            /* Enable error reporting for expressions and declarations with an implied 'any' type. */
    // "strictNullChecks": true,                         /* When type checking, take into account 'null' and 'undefined'. */
    // "strictFunctionTypes": true,                      /* When assigning functions, check to ensure parameters and the return values are subtype-compatible. */
    // "strictBindCallApply": true,                      /* Check that the arguments for 'bind', 'call', and 'apply' methods match the original function. */
    // "strictPropertyInitialization": true,             /* Check for class properties that are declared but not set in the constructor. */
    // "noImplicitThis": true,                           /* Enable error reporting when 'this' is given the type 'any'. */
    // "useUnknownInCatchVariables": true,               /* Default catch clause variables as 'unknown' instead of 'any'. */
    // "alwaysStrict": true,                             /* Ensure 'use strict' is always emitted. */
    // "noUnusedLocals": true,                           /* Enable error reporting when local variables aren't read. */
    // "noUnusedParameters": true,                       /* Raise an error when a function parameter isn't read. */
    // "exactOptionalPropertyTypes": true,               /* Interpret optional property types as written, rather than adding 'undefined'. */
    // "noImplicitReturns": true,                        /* Enable error reporting for codepaths that do not explicitly return in a function. */
    // "noFallthroughCasesInSwitch": true,               /* Enable error reporting for fallthrough cases in switch statements. */
    // "noUncheckedIndexedAccess": true,                 /* Add 'undefined' to a type when accessed using an index. */
    // "noImplicitOverride": true,                       /* Ensure overriding members in derived classes are marked with an override modifier. */
    // "noPropertyAccessFromIndexSignature": true,       /* Enforces using indexed accessors for keys declared using an indexed type. */
    // "allowUnusedLabels": true,                        /* Disable error reporting for unused labels. */
    // "allowUnreachableCode": true,                     /* Disable error reporting for unreachable code. */

    /* Completeness */
    // "skipDefaultLibCheck": true,                      /* Skip type checking .d.ts files that are included with TypeScript. */
    "skipLibCheck": true                                 /* Skip type checking all .d.ts files. */
  }
}

```

**Comment by warcayac** [comment]
in my case I just had to enable **lib** key and set this:
```
    "lib": ["ESNext"],
```

**Comment by maranomynet** [comment]
ok, I realized I have to add `"lib": ["ESNext", "DOM"]` to my `./tsconfig.file` — since I'm writing code that interact with the DOM.
And, BOOM! my old fiend rears its ugly head:

```
Property 'json' does not exist on type 'BunFile'.
```

**Comment by maranomynet** [comment]
I'm confused about what's the correct way to write browser-code in TypeScript in bun.
Do I need to set up multiple tsconfig.json files for all such projects?

**Comment by warcayac** [comment]
> > I believe you have `"DOM"` in `tsconfig.json`s `lib` configuration.
> 
> I don't think so. Here's my `tsconfig.json`:
> 
> ```js
> {
>   "compilerOptions": {
>     "lib": ["esnext"],
>     "allowSyntheticDefaultImports": true,
>     "module": "NodeNext",
>     "target": "ES2017",
>     "esModuleInterop": true,
>     "resolveJsonModule": true,
>     "skipLibCheck": true,
>     "noEmit": true,
>     "jsx": "react",
> 
>     "strict": true.
>     "noUncheckedIndexedAccess": true,
>     "noImplicitAny": true,
>     "useUnknownInCatchVariables": true
>   },
>   // include and exclude ...
> }
> ```

try to change to these values:
```json
"target": "ES2022",
"lib": ["es2022", "dom", "dom.iterable"],
"module": "ES2022",
"moduleResolution": "Bundler",
```
if you are not running in the DOM then: `"lib": [],`

**Comment by maranomynet** [comment]
@warcayac It has no effect...

<img width="580" alt="image" src="https://github.com/oven-sh/bun/assets/86827/72d81718-e354-4cd2-9b59-5a496723ce6b">

**Comment by warcayac** [comment]
your original code is similar to mine, so the settings I suggested work fine for me, for the last code, do you have the same error? why don't you separate them into variables to debug?

**Comment by maranomynet** [comment]
Ah, yes it is the same error.
<img width="902" alt="image" src="https://github.com/oven-sh/bun/assets/86827/9fe0bd79-71e1-4da6-9850-53bc01355e4f">

<img width="365" alt="image" src="https://github.com/oven-sh/bun/assets/86827/b92c4f2c-fd02-4237-9240-f7babcd5d00e">

Separated:
<img width="647" alt="image" src="https://github.com/oven-sh/bun/assets/86827/0df8d457-3c3d-4c13-9c32-1c3c0033cefa">

**Comment by maranomynet** [comment]
BTW: Before taking the above screenshots I did a clean install (see list of dependencies):

```
mar:~/work/lib-starter$ rm -rf node_modules/ bun.lockb 
mar:~/work/lib-starter$ bun install
bun install v1.0.23 (83f2432d)

 + @types/bun@1.0.1
 + @types/glob@8.1.0
 + glob@10.3.10
 + typescript@4.9.5 (v5.3.3 available)

 43 packages installed [38.00ms]
 ```

**Comment by warcayac** [comment]
I'm gonna try reproduce your error, can you send me the `./tsconfig.build.json` if it doesn't contain sensitive info. Moreover can you tell if your App is a server or frontend, if you use Vite or Elysia, etc ?

**Comment by maranomynet** [comment]
Here's a minimal reproduction:
[bun-replicate--issue-8103.zip](https://github.com/oven-sh/bun/files/13961508/bun-replicate--issue-8103.zip)

<img width="511" alt="image" src="https://github.com/oven-sh/bun/assets/86827/874695e7-edf4-4b7a-bdc3-0975d320bc64">

The repro project was created via these steps, with bun v1.0.23

1. `bun init -y`
2. Edit `tsconfig.json` to add `"DOM"` to `tsCfg.compilerOptions.lib`
3. Edit `index.ts` to insert line of code that produces the type error

**BTW:**
```
$ grep "version" node_modules/\@types/bun/package.json 
    "version": "1.0.2",
```

**Comment by warcayac** [comment]
Hey @maranomynet ,

Good news: your problem is reproducible
Bad news: no solution. I've tried with different configurations of tsconfig.json 

Bun doc helps little, I've found this reference only: https://bun.sh/docs/test/dom

If you are trying to scrape a web page with Bun and TS, see this page: 
https://apify.com/templates/ts-start-bun

My previous suggestion was based on this page: https://www.totaltypescript.com/tsconfig-cheat-sheet
but I didn't need to use "dom" and "dom.iterable"

**Comment by maranomynet** [comment]
I'm just trying to use bun for normal things I'd normally do with node or yarn — like develop a npm library that makes use of d3 (svg charts).

Bun's types feel broken, and this breakage started with @types/bun@1.0.1

**Comment by lgarron** [comment]
I'm also running into this. I have a project with containing pretty much just `bun-types` at `v1.0.24` and it's running into this:

![Screenshot 2024-01-20 at 16 08 05](https://github.com/oven-sh/bun/assets/248078/7a61866e-3f87-4bb3-899b-09b208a2dd9b)

**Comment by lgarron** [comment]
Okay, figured out a few issues while fixing this:

- `bun init` tries to add `@bun/types` instead of `bun-types`, but that's out of date: https://github.com/oven-sh/bun/issues/8301
- `bun-types` needs to be explicitly included in e.g. the `"types"` field.

**Comment by maranomynet** [comment]
Happy to report this seems to be fixed, at least as of `bun` v1.1.17 and `@types/bun` v1.1.6

```sh
bun-replicate--issue-8103$ rm bun.lockb 
bun-replicate--issue-8103$ bun install
bun install v1.1.17 (bb66bba1)

+ typescript@5.5.3
+ @types/bun@1.1.6

5 packages installed [2.14s]
bun-replicate--issue-8103$ 
```

opening `index.ts` in VSCode no longer displays a type error.

<img width="340" alt="image" src="https://github.com/oven-sh/bun/assets/86827/b105ed1d-baf1-4855-bc46-e619bae00177">
